scripts/apps/authoring/authoring/directives/ItemCarouselDirective.ts

```diff
@@ -49,17 +50,18 @@ export function ItemCarouselDirective($timeout, notify) {
              * otherwise carousel height is messed up
              */
             scope.$watch('items', (items) => {
-                // Don't execute if there are no items or their length is same as before
-                if (items == null || items.length === previousItemsLength) {
+                let field = _.find(items, (item) => !item[item.fieldId]);
+
+                scope.rel = field ? field.fieldId : null;
+
+                // Don't execute if there are no items or their length is same as before and
+                // previousRel (previous next related-key of media item) is same as current rel(next key of media item)
+                if (items == null || items.length === previousItemsLength && previousRel === scope.rel) {
```

Comment:
Why not use lodash `isEqual(items, scope.items)`?

Reply:
Actually, `items` and `scope.items` both are same, as we are watching `scope.iems` and `items` is the current value of `scope.items`.

Reply:
Isn't `items` received in a callback from `watch` is an updated version which you could check against the current state in order to decide if you want to update the view?

Reply:
> items received in a callback from a watch is an updated version.

Yep, I do agree that `items` is an updated version, but it could be used to check against the previous state, not current state, as a watch is triggered only after the current state is changed.
![watch-scope](https://user-images.githubusercontent.com/10600741/62684913-49e37600-b9df-11e9-873f-641a46bce7f1.png)

Reply:
When is `previousRel === scope.rel` check relevant? What is it checking?

Reply:
And the thing is we have to reload the carousel only if an item is removed or added, otherwise on every single change carousel will reload and the cursor will be lost. So, it's better to check the length than the whole content.

Reply:
> When is previousRel === scope.rel check relevant? What is it checking?

As I understood `scope.rel` holds the next key of the media that item, let's say we can add 3 items at max in a media gallery field, and we have already added two items with key `media--1, media--2`, then  `scope.rel` will be `scope.rel = media--3`. And when we try to add the third item `scope.rel` would become `null`. So, `previousRel === scope.rel` actually signifies that length is changed or not. So, I think there is no need to check the length of the items as well.

Reply:
> And when we try to add the third item scope.rel would become null. So, previousRel === scope.rel actually signifies that length is changed or not. So, I think there is no need to check the length of the items as well.

These are side-effects. Check what you need directly. You mentioned that carousel needs to be reloaded when items are added or removed - check for that directly:

```
function getItemsCount(items) {
    return Object.keys(items)
        .filter((key) => items[key] != null)
        .length;
}

if (getItemsCount(previousItems) !== getItemsCount(currentItems)) {
    // reload carousel
}
```